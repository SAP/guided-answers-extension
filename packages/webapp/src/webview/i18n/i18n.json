{
    "GUIDED_ANSWER_HEADER": "Guide",
    "SEARCH_GUIDED_ANSWERS": "Search Guided Answers",
    "ALL_ANSWERS": "Home",
    "STEP_BACK": "Step back",
    "RESTART": "Restart",
    "GUIDED_ANSWERS": "Guided Answers",
    "NO_ANSWERS_FOUND": "No answers found",
    "PLEASE_MODIFY_SEARCH": "Please modfiy search",
    "THIS_SOLVED_MY_ISSUE": "This solved my issue",
    "THIS_DID_NOT_SOLVED_MY_ISSUE": "This did not solve my issue",
    "THANK_YOU_TEXT": "We are glad to hear that your issue has been resolved and we hope you enjoy using the Guided Answers application.",
    "THANKS": "Thanks!",
    "HOME": "Home",
    "SEND": "Send",
    "CLOSE": "Close",
    "ISSUE_IS_NOT_RESOLVED": "Issue is not resolved",
    "WE_ARE_SORRY_TO_HEAR_THAT_YOUR_ISSUE_IS_NOT_YET_RESOLVED": "We are sorry to hear that your issue is not yet resolved.",
    "PLEASE_TELL_US_IF_THIS_ANSWER_WAS_HELPFUL": "Please tell us if this answer was helpful",
    "THERE_ARE_SEVERAL_OPTIONS_FOR_GETTING_FURTHER_ASSISTANCE": "There are several options for getting further assistance:",
    "SEARCH_FOR_ANOTHER_GUIDED_ANSWER": "Search for another Guided Answer",
    "START_AN_EXPERT_CHAT": "Start an Expert Chat",
    "SCHUEDULE_AN_EXPERT": "Schedule an Expert",
    "OPEN_AN_INCIDENT": "Open an Incident",
    "ASK_THE_SAP_COMMUNITY": "Ask the SAP Community",
    "FEEDBACK": "Feedback",
    "FEEDBACK_DIALOG_TITLE": "Is this content helpful?",
    "FEEDBACK_DIALOG_SUBTEXT": "If you have suggestions on how to improve this content we would love to hear them!",
    "FEEDBACK_DIALOG_DISCLAIMER": "Your feedback is anonymous, we do not collect any personal data.",
    "FEEDBACK_DIALOG_SUGGESTION": "Your suggestion",
    "FEEDBACK_SENT_DIALOG_TITLE": "Message sent",
    "FEEDBACK_SENT_DIALOG_SUBTEXT": "Thanks for the feedback!"
}
