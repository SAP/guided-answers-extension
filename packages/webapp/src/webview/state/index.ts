export { actions, store } from './store';
export { getInitialState } from './reducers';
