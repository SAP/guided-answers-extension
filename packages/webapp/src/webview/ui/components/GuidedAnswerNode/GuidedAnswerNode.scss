.answer__node__body {
    height: 100%;
    margin: 0px;
    display: flex;
    box-sizing: border-box;
}

.body_container {
    overflow: auto;
    overflow-wrap: anywhere;
}

header {
    font-weight: 700;
    font-size: 24px;
    line-height: 29px;
}

button {
    background: none;
    color: inherit;
    border: none;
    padding: 0;
    font: inherit;
    cursor: pointer;
    outline: inherit;
}

button:focus-visible {
    outline: 1px solid var(--vscode-focusBorder, #007fd4);
}

.column {
    flex-direction: column;
    box-sizing: border-box;
    justify-content: flex-start;
    align-items: stretch;
    flex-wrap: nowrap;
}

a:hover {
    outline: 1px solid var(--vscode-contrastActiveBorder);
}

#left {
    height: 100%;
    width: 25%;
    margin-right: 80px;
}

.main-container {
    display: flex;
}

#middle {
    display: flex;
    max-width: 620px;
    width: 75%;
    align-content: flex-start;

    .content > * {
        line-height: 18px;
    }

    .content > h1 {
        font-size: 20px;
        margin-top: 15px;
    }

    a:focus-visible {
        outline: 1px solid var(--vscode-focusBorder, #007fd4);
    }
}

#hr {
    width: 100%;
    border-bottom: 1px solid var(--vscode-editorWidget-border);
    padding-bottom: 10px;
}

#middle-with-command {
    width: 370vw;
    display: flex;
    grid-area: content;
    overflow-x: hidden;
    position: relative;
    left: 0;
    transition: left 0.5s;
    flex-grow: 1;
    margin-left: 400px;
}

#right {
    min-width: 310px;
    padding-top: 32px;
}

.bottom {
    flex-grow: 1;
    overflow-y: auto;
}

.guided-answer__node {
    & {
        display: flex;
        flex-direction: column;
    }
    &__body {
        color: var(--vscode-foreground);
        display: flex;
    }
    &__question {
        color: var(--vscode-settings-headerForeground);
        font-size: 14px;
        font-style: normal;
        font-weight: 700;
        line-height: 16px;
        letter-spacing: 0em;
        text-align: left;
    }
    &__edge {
        background-color: var(--vscode-button-background);
        color: var(--vscode-button-foreground);
        cursor: pointer;
        font-size: 13px;
        font-style: normal;
        font-weight: 700;
        height: auto;
        line-height: 18px;
        letter-spacing: 0em;
        margin: 10px 0;
        padding: 10px;
        text-align: center;
        outline-offset: -1px;
        border: 1px solid var(--vscode-contrastBorder);
        text-decoration: none;
    }
    &__edge:hover {
        background-color: var(--vscode-button-hoverBackground);
        outline: 1px solid var(--vscode-contrastActiveBorder);
    }
    &__edge:focus-visible {
        outline: 1px solid var(--vscode-focusBorder);
        outline-offset: 2px;
    }

    &__commands {
        position: fixed;
        margin-top: 18px;
    }
    &__command {
        margin: 16px;
        background-color: var(--vscode-sideBar-background);
        border-radius: 8px;
        padding: 10px;
        &__header {
            text-align: left;
            display: flex;
            &__icon {
                width: 16px;
                height: 16px;
                padding-top: 2px;
                vertical-align: bottom;
            }
            &__label {
                color: var(--vscode-foreground);
                font-size: 14px;
                font-weight: 700;
                line-height: 18px;
                padding-left: 5px;
                white-space: nowrap;
                display: inline-block;
            }
        }
        &__description {
            margin: 7px;
            position: relative;
            color: var(--vscode-foreground);
            font-size: 13px;
            font-weight: 400;
            line-height: 18px;
            height: 50%;
            text-align: start;
        }
    }
    &__command:hover {
        background-color: var(--vscode-list-hoverBackground);
        outline: 1px solid var(--vscode-contrastActiveBorder);
        cursor: pointer;
    }
}

.enhancement-link {
    color: var(--vscode-textLink-foreground);
    cursor: pointer;
    text-decoration: underline;
}

.enhancement-link:focus-visible {
    outline: 1px solid var(--vscode-focusBorder, #007fd4);
}

a:hover {
    color: var(--vscode-button-foreground);
}

@media only screen and (max-width: 625px) {
    #middle {
        width: 100%;
    }

    #left {
        width: auto;
    }
}
