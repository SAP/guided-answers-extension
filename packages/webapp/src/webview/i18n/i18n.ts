import { use } from 'i18next';
import { initReactI18next } from 'react-i18next';
import * as i18nEn from './i18n.json';

export { i18n } from 'i18next';

export function initI18n(): void {
    use(initReactI18next)
        .init({
            resources: {
                en: {
                    translation: i18nEn
                }
            },
            lng: 'en',
            fallbackLng: 'en',
            interpolation: {
                escapeValue: false
            }
        })
        .catch((error) => console.error(error));
}
