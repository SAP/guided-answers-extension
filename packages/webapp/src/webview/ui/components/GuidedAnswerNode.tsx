import { default as parse, DOMNode, Element, Text } from 'html-react-parser';
import React, { ReactElement } from 'react';
import { GuidedAnswerNode, HTML_ENHANCEMENT_DATA_ATTR_MARKER } from '@sap/guided-answers-extension-types';
import { useSelector } from 'react-redux';
import { actions } from '../../state';
import { AppState } from '../../types';
import './GuidedAnswerNode.scss';
import { GuidedAnswerNavPath } from './GuidedAnswerNavPath';

/**
 * Replacer function for html-react-parser's replace function. If an element was marked, replace it with  link <a>
 * The command JSON is in the data-* attribute.
 *
 * @param domNode - current DOM node from html-react-parser
 * @returns - undefined if nothing to replace; the new node (<a>) in case of replacement
 */
function replace(domNode: DOMNode): JSX.Element | undefined {
    if (domNode.type === 'tag') {
        const domElement: Element = domNode as Element;
        const dataCommandString = domElement?.attribs?.[HTML_ENHANCEMENT_DATA_ATTR_MARKER];
        if (dataCommandString) {
            try {
                const command = JSON.parse(decodeURIComponent(dataCommandString));
                const textContent = domElement?.firstChild?.type === 'text' ? (domElement.firstChild as Text).data : '';
                if (command) {
                    return (
                        <a
                            href="javascript:void(0)"
                            onClick={(): void => {
                                actions.executeCommand(command);
                            }}>
                            {textContent}
                        </a>
                    );
                }
            } catch (error) {
                console.error(error);
            }
        }
    }
}

/**
 * Process the enhancements in an HTML string, which includes parsing the HTML and replacing the respective nodes.
 *
 * @param htmlString - HTML string
 * @returns - react elements including enhancements
 */
function enhanceBodyHtml(htmlString: string): ReactElement | undefined {
    let result;
    try {
        result = parse(htmlString, { replace });
    } catch (error) {
        console.error(error);
    }
    return result as ReactElement;
}

/**
 * Check if an HTML string contains an enhancement.
 *
 * @param htmlString - HTML string
 * @returns true: has enhancementsl; false: no enhancements
 */
function hasEnhancements(htmlString: string): boolean {
    return typeof htmlString === 'string' && htmlString.includes(HTML_ENHANCEMENT_DATA_ATTR_MARKER);
}

/**
 * Return the navigation section.
 *
 * @returns - react elements for navigation
 */
function getNavigationSection(): ReactElement {
    return (
        <div id="left" className="column">
            <GuidedAnswerNavPath />
        </div>
    );
}

/**
 * Return the content for a Guided Answers node, which consists of the middle part, the actual Guided Answers text
 * and, if present, the right part with additional commands.
 *
 * @param activeNode - the active Guided Answers node
 * @returns - react element for content
 */
function getContent(activeNode: GuidedAnswerNode): ReactElement {
    const enhancedBody = hasEnhancements(activeNode.BODY) ? enhanceBodyHtml(activeNode.BODY) : null;
    const middle = (
        <div id="middle" className="column">
            <h1>
                {activeNode.TITLE} <span className="light-text">#{activeNode.NODE_ID}</span>
            </h1>
            {enhancedBody ? enhancedBody : <div dangerouslySetInnerHTML={{ __html: activeNode.BODY }}></div>}
            <p className="guided-answer__node__question">{activeNode.QUESTION}</p>
            <div className="guided-answer__node">
                {activeNode.EDGES.map((edge, index) => (
                    <div
                        key={`edge_button${index}`}
                        className="guided-answer__node__edge"
                        onClick={(): void => {
                            actions.selectNode(edge.TARGET_NODE);
                        }}>
                        {edge.LABEL}
                    </div>
                ))}
            </div>
        </div>
    );
    const right = activeNode.COMMANDS ? (
        <div id="right" className="column">
            <div className="guided-answer__node__commands">
                {activeNode.COMMANDS
                    ? activeNode.COMMANDS.map((command, index) => (
                          <div className="guided-answer__node__command">
                              <div className="guided-answer__node__command__header">
                                  {/* <div className="guided-answer__node__command__header__icon">
                              {command.icon}
                          </div> */}
                                  <div className="guided-answer__node__command__header__label">{command.label}</div>
                              </div>

                              <div
                                  key={`command-${index}`}
                                  className="guided-answer__node__command__description"
                                  onClick={(): void => {
                                      actions.executeCommand(command);
                                  }}>
                                  {command.description}
                              </div>
                          </div>
                      ))
                    : ''}
            </div>
        </div>
    ) : null;

    return right ? (
        <div className="main-container">
            {middle}
            {right}
        </div>
    ) : (
        middle
    );
}

/**
 * Render the react elements to display a Guided Answers node.
 *
 * @returns - react element of Guided Answers node
 */
export function GuidedAnswerNode(): ReactElement {
    const nodes = useSelector<AppState, GuidedAnswerNode[]>((state) => state.activeGuidedAnswerNode);
    const activeNode = nodes[nodes.length - 1];
    return activeNode ? (
        <section className="guided-answer__node__body">
            {getNavigationSection()}
            {getContent(activeNode)}
        </section>
    ) : (
        <></>
    );
}
