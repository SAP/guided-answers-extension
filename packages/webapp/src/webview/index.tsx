import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { App } from './ui/components/App';
import { actions, store } from './state';
import { initI18n } from './i18n';

/**
 * Initialization of i18n, icons
 */
initI18n();

/**
 * Render app
 */
render(
    <Provider store={store}>
        <App />
    </Provider>,
    document.getElementById('root') as HTMLElement
);

/**
 * Fire action which is captured in hosting app (ide-extension) to signal that we are ready to receive messages
 */
actions.webviewReady();

/**Once webview is ready we focus to search field */
document.getElementById('search-field')?.focus();
