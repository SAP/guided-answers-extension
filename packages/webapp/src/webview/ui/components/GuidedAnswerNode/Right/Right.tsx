import React, { ReactElement } from 'react';
import { actions } from '../../../../state';
import type { GuidedAnswerNode as GuidedAnswerNodeType } from '@sap/guided-answers-extension-types';
import '../GuidedAnswerNode.scss';
import { FocusZone, FocusZoneDirection } from '@fluentui/react-focus';

/**
 * @param props - props for middle component
 * @param props.activeNode - array that stores GuidedAnswers node objects
 * @returns - The middle react element
 */
export function Right(props: { activeNode: GuidedAnswerNodeType }): ReactElement {
    return (
        <div id="right" className="column">
            <div className="guided-answer__node__commands">
                <FocusZone direction={FocusZoneDirection.vertical} isCircularNavigation={true}>
                    {props.activeNode.COMMANDS
                        ? props.activeNode.COMMANDS.map((command, index) => (
                              <div className="guided-answer__node__command" key={`command-${index}`}>
                                  <div className="guided-answer__node__command__header">
                                      <div className="guided-answer__node__command__header__label">{command.label}</div>
                                  </div>

                                  <button
                                      className="guided-answer__node__command__description"
                                      onClick={(): void => {
                                          actions.executeCommand(command);
                                      }}>
                                      {command.description}
                                  </button>
                              </div>
                          ))
                        : ''}
                </FocusZone>
            </div>
        </div>
    );
}
