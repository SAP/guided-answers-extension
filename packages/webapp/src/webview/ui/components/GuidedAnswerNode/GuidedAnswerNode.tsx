import { default as parse, DOMNode, Element, Text } from 'html-react-parser';
import React, { ReactElement } from 'react';
import type { Command, GuidedAnswerNode as GuidedAnswerNodeType } from '@sap/guided-answers-extension-types';
import { HTML_ENHANCEMENT_DATA_ATTR_MARKER } from '@sap/guided-answers-extension-types';
import { useSelector } from 'react-redux';
import { actions } from '../../../state';
import { AppState } from '../../../types';
import './GuidedAnswerNode.scss';
import { GuidedAnswerNavPath } from '../GuidedAnswerNavPath';
import { Middle } from './Middle';
import { Right } from './Right';

/**
 * Replacer function for html-react-parser's replace function. If an element was marked, replace it with  link <a>
 * The command JSON is in the data-* attribute.
 *
 * @param domNode - current DOM node from html-react-parser
 * @returns - undefined if nothing to replace; the new node (<a>) in case of replacement
 */
function replace(domNode: DOMNode): ReactElement | undefined {
    let result: ReactElement | undefined;
    if (domNode.type === 'tag') {
        const domElement: Element = domNode as Element;
        const dataCommandString = domElement?.attribs?.[HTML_ENHANCEMENT_DATA_ATTR_MARKER];
        if (dataCommandString) {
            try {
                const command = JSON.parse(decodeURIComponent(dataCommandString)) as Command;
                const textContent = domElement?.firstChild?.type === 'text' ? (domElement.firstChild as Text).data : '';
                if (command) {
                    result = (
                        <button
                            title={command.description}
                            className="enhancement-link"
                            onClick={(): void => {
                                actions.executeCommand(command);
                            }}>
                            {textContent}
                        </button>
                    );
                }
            } catch (error) {
                console.error(error);
            }
        }
    }
    return result;
}

/**
 * Process the enhancements in an HTML string, which includes parsing the HTML and replacing the respective nodes.
 *
 * @param htmlString - HTML string
 * @returns - react elements including enhancements
 */
function enhanceBodyHtml(htmlString: string): ReactElement | undefined {
    let result;
    try {
        result = parse(htmlString, { replace });
    } catch (error) {
        console.error(error);
    }
    return result as ReactElement;
}

/**
 * Check if an HTML string contains an enhancement.
 *
 * @param htmlString - HTML string
 * @returns true: has enhancementsl; false: no enhancements
 */
function hasEnhancements(htmlString: string): boolean {
    return typeof htmlString === 'string' && htmlString.includes(HTML_ENHANCEMENT_DATA_ATTR_MARKER);
}

/**
 * Return the navigation section.
 *
 * @returns - react elements for navigation
 */
function getNavigationSection(): ReactElement {
    return (
        <div id="left" className="column">
            <GuidedAnswerNavPath />
        </div>
    );
}

/**
 * Return the content for a Guided Answers node, which consists of the middle part, the actual Guided Answers text
 * and, if present, the right part with additional commands.
 *
 * @param activeNode - the active Guided Answers node
 * @returns - react element for content
 */
function getContent(activeNode: GuidedAnswerNodeType): ReactElement {
    const enhancedBody = hasEnhancements(activeNode.BODY) ? enhanceBodyHtml(activeNode.BODY) : null;
    if (activeNode.COMMANDS) {
        return (
            <div className="main-container">
                {<Middle activeNode={activeNode} enhancedBody={enhancedBody} />}
                {<Right activeNode={activeNode} />}
            </div>
        );
    } else {
        return <Middle activeNode={activeNode} enhancedBody={enhancedBody} />;
    }
}

/**
 * Render the react elements to display a Guided Answers node.
 *
 * @returns - react element of Guided Answers node
 */
export function GuidedAnswerNode(): ReactElement {
    const nodes = useSelector<AppState, GuidedAnswerNodeType[]>((state) => state.activeGuidedAnswerNode);
    const activeNode = nodes[nodes.length - 1];
    return activeNode ? (
        <section className="guided-answer__node__body">
            {getNavigationSection()}
            {getContent(activeNode)}
        </section>
    ) : (
        <></>
    );
}
