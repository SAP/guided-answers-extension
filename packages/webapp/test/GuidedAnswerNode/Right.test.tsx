import React from 'react';
import { shallow } from 'enzyme';
import { Right } from '../../src/webview/ui/components/GuidedAnswerNode/Right';
import { actions } from '../../src/webview/state';
import { initI18n } from '../../src/webview/i18n';

jest.mock('../../src/webview/state', () => {
    return {
        actions: {
            executeCommand: jest.fn()
        }
    };
});

describe('<Right />', () => {
    let wrapper: any;
    initI18n();
    const activeGuidedAnswerNodeMock = [
        {
            BODY: '<p>SAP Fiori Tools is a set of extensions for SAP Business Application Studio and Visual Studio Code</p>',
            EDGES: [
                { LABEL: 'Deployment', TARGET_NODE: 45996, ORD: 1 },
                { LABEL: 'Fiori Generator', TARGET_NODE: 48363, ORD: 2 }
            ],
            NODE_ID: 45995,
            QUESTION: 'I have a problem with',
            TITLE: 'SAP Fiori Tools'
        }
    ];

    afterEach(() => {
        jest.clearAllMocks();
        wrapper.unmount();
    });

    it('Should render a Right component without command', () => {
        wrapper = shallow(<Right activeNode={activeGuidedAnswerNodeMock[0]} />);
        const component = wrapper.html();
        expect(component).toMatchInlineSnapshot(
            `"<div id="right" class="column"><div class="guided-answer__node__commands"><div class="ms-FocusZone css-101" data-focuszone-id="FocusZone0"></div></div></div>"`
        );
    });

    it('Should render a Right component', () => {
        const activeNodeMock: any = activeGuidedAnswerNodeMock[0];
        activeNodeMock.COMMANDS = [
            {
                label: 'Label for command',
                description: 'Description for command',
                exec: {
                    extensionId: 'terry.exxt',
                    commandId: 'Knock kock',
                    argument: { fsPath: 'whos/there/body/of' }
                }
            }
        ];
        wrapper = shallow(<Right activeNode={activeNodeMock} />);
        const component = wrapper.html();
        expect(component).toMatchInlineSnapshot(
            `"<div id="right" class="column"><div class="guided-answer__node__commands"><div class="ms-FocusZone css-101" data-focuszone-id="FocusZone1"><div class="guided-answer__node__command"><div class="guided-answer__node__command__header"><div class="guided-answer__node__command__header__label">Label for command</div></div><button class="guided-answer__node__command__description">Description for command</button></div></div></div></div>"`
        );

        //Test click event
        wrapper.find('.guided-answer__node__command__description').simulate('click');
        expect(actions.executeCommand).toBeCalled();
    });
});
