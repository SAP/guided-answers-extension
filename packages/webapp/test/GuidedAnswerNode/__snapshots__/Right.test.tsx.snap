// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Right /> Should render a Right component 1`] = `
<div>
  <div
    class="guided-answer__node__commands"
  >
    <div
      class="ms-FocusZone css-166"
      data-focuszone-id="FocusZone1"
    >
      <div
        class="guided-answer__node__command"
        role="button"
        title="Label for command"
      >
        <div
          class="guided-answer__node__command__header"
        >
          <i
            aria-hidden="true"
            class="ts-icon guided-answer__node__command__header__icon root-162"
            data-icon-name="CreateMockData"
          />
          <div
            class="guided-answer__node__command__header__label"
          >
            Label for command
          </div>
        </div>
        <button
          class="guided-answer__node__command__description"
          id="guided-answer__node__command"
          tabindex="0"
        >
          Description for command
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Right /> Should render a Right component without command 1`] = `
<div>
  <div
    class="guided-answer__node__commands"
  >
    <div
      class="ms-FocusZone css-166"
      data-focuszone-id="FocusZone0"
    />
  </div>
</div>
`;
